crates/nu-command/src/filters/chunk_by.rs

```diff
@@ -0,0 +1,220 @@
+    iterator: I,
+    last_value: Option<(T, K)>,
+    closure: F,
+    done: bool,
+}
+
+impl<I, T, F, K> Iterator for Chunk<I, T, F, K>
+where
+    I: Iterator<Item = T>,
+    F: FnMut(&T) -> K,
+    K: PartialEq,
+{
+    type Item = Vec<T>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.done {
+            return None;
+        }
+
+        let (head, head_key) = match self.last_value.take() {
+            None => {
+                let head = self.iterator.next()?;
+
+                let key = (self.closure)(&head);
+
+                (head, key)
+            }
+
+            Some((value, key)) => (value, key),
+        };
+
+        let mut result = vec![head];
+
+        loop {
+            match self.iterator.next() {
+                None => {
+                    self.done = true;
+                    return Some(result);
+                }
+                Some(value) => {
+                    let value_key = (self.closure)(&value);
+
+                    if value_key == head_key {
+                        result.push(value);
+                    } else {
+                        self.last_value = Some((value, value_key));
+                        return Some(result);
+                    }
+                }
+            }
+        }
+    }
+}
+
+/// An iterator with the semantics of the chunk_by operation.
+fn chunk_iter_by<I, T, F, K>(iterator: I, closure: F) -> Chunk<I, T, F, K>
+where
+    I: Iterator<Item = T>,
+    F: FnMut(&T) -> K,
+    K: PartialEq,
+{
+    Chunk {
+        closure,
+        iterator,
+        last_value: None,
+        done: false,
+    }
+}
+
+pub fn chunk_by(
+    engine_state: &EngineState,
+    stack: &mut Stack,
+    call: &Call,
+    input: PipelineData,
+) -> Result<PipelineData, ShellError> {
+    let head = call.head;
+    let closure: Closure = call.req(engine_state, stack, 0)?;
+
+    let metadata = input.metadata();
+
+    match input {
+        PipelineData::Empty => Ok(PipelineData::Empty),
+        PipelineData::Value(Value::Range { .. }, ..)
+        | PipelineData::Value(Value::List { .. }, ..)
+        | PipelineData::ListStream(..) => {
+            let closure = ClosureEval::new(engine_state, stack, closure);
+
+            let result = chunk_value_stream(input.into_iter(), closure, head);
+
+            Ok(result.into_pipeline_data(head, engine_state.signals().clone()))
+        }
+        PipelineData::ByteStream(stream, ..) => {
+            if let Some(chunks) = stream.chunks() {
+                let closure = ClosureEval::new(engine_state, stack, closure);
+
+                let mapped_chunks =
+                    chunks.map(move |value| value.unwrap_or_else(|err| Value::error(err, head)));
+
+                let result = chunk_value_stream(mapped_chunks, closure, head);
+
+                Ok(result.into_pipeline_data(head, engine_state.signals().clone()))
+            } else {
+                Ok(PipelineData::Empty)
+            }
+        }
```

Comment:
I think this should return a type error. `ByteStream`s are of type `string` or `binary`, but `chunk-by` only operates on `list`s.

Reply:
Oh I had forgotten about that. your suggestion makes sense